[Konvoy 2.2] Upgrade Docs (#4230)
mesosphere/dcos-docs-site: pages/dkp/kommander/2.2/dkp-upgrade/upgrade-konvoy/enterprise/index.md

---
layout: layout.pug
navigationTitle: DKP Enterprise
title: Upgrade Konvoy for DKP Enterprise
excerpt: Upgrade your Konvoy environment within the DKP Enterprise license.
beta: false
enterprise: false
menuWeight: 30
---

## Prerequisites

> * Follow the steps listed in the [DKP upgrade overview][dkpup]

Reviewer: This link doesn't lead to anywehre

Author: Odd, i'll edit.

> * Ensure that all platform applications in the management cluster have been upgraded to avoid compatibility issues with the current Kubernetes version. This is done automatically when [upgrading Kommander][upgradekomm], so ensure that Kommander has been upgraded prior to Konvoy.

Reviewer 2: ```suggestion
* Ensure that all Platform applications in the Management cluster have been upgraded to avoid compatibility issues with the [Kubernetes version][releasenotes] included in this release. This is done automatically when [upgrading Kommander][upgradekomm], so ensure that Kommander has been upgraded prior to Konvoy.
```

> * For air-gapped: Download the required bundles either at our [support site][supportsite] or [via the CLI][airgapbundle]

Reviewer 2: ```suggestion
* For air-gapped: Download the required bundles either at our [support site][supportsite] or [via the CLI][airgapbundle].
```

The following infrastructure environments are supported:

* Amazon Web Services (AWS)

* Microsoft Azure

* Pre-provisioned environments

## Overview

The Konvoy upgrade for DKP Enterprise encompasses the following three steps in sequential order:

1. Upgrade the Cluster API (CAPI) components
1. Upgrade the core addons
1. Upgrade the Kubernetes version

Run all three steps on the management cluster (Kommander cluster) first. Then, run the second and third step on additional managed clusters (Konvoy clusters), one cluster at the time.

For a full list of DKP Enterprise features, see [DKP Enterprise][dkpenterprise].

<p class="message--note"><strong>NOTE:</strong> You must maintain your attached clusters manually. Review the documentation from your cloud provider for more information.

## Upgrade the CAPI components

New versions of DKP come pre-bundled with newer versions of CAPI, newer versions of infrastructure providers or new infrastructure providers. When using a new version of the DKP CLI, upgrade all of these components first.

> If you are running on more than one management cluster (Kubernetes cluster provisioned with DKP), you must upgrade the CAPI Components on each of these clusters. Ensure your `dkp` configuration references the management cluster where you want to run the upgrade by setting the KUBECONFIG environment variable [to the appropriate kubeconfig file’s location][kubeconfig].

Reviewer: ```suggestion
If you are running on more than one management cluster (Kubernetes cluster provisioned with DKP), you must upgrade the CAPI Components on each of these clusters. Ensure your `dkp` configuration references the management cluster where you want to run the upgrade by setting the `KUBECONFIG` environment variable [to the appropriate kubeconfig file’s location][kubeconfig].
```

<p class="message--note"><strong>NOTE:</strong> An alternative to initializing the KUBECONFIG environment variable is to use the <code>--kubeconfig=cluster_name.conf</code> flag, ensuring that Kommander upgrades on the workload cluster.</p>

> 1. Run the following upgrade command for the CAPI components.

Reviewer: We need to mention that users need to set the appropriate envs for AWS and Azure.

For Azure they should always set envs like in  http://docs-d2iq-com-pr-4230.s3-website-us-west-2.amazonaws.com/dkp/konvoy/2.2/choose-infrastructure/azure/quick-start-azure/#configure-azure-prerequisites

For AWS it would be http://docs-d2iq-com-pr-4230.s3-website-us-west-2.amazonaws.com/dkp/konvoy/2.2/choose-infrastructure/aws/quick-start-aws/#configure-aws-prerequisites

Author: So would this be a prereq before starting any of the steps?

```bash
dkp upgrade capi-components
```

The command should output something similar to the following:

```bash
✓ Upgrading CAPI components
✓ Waiting for CAPI components to be upgraded
✓ Initializing new CAPI components
✓ Deleting Outdated Global ClusterResourceSets
```

1. Once finished, verify that the components were upgraded using the `kubectl get provider` command.

```bash

> kubectl get provider --all-namespaces

Reviewer: This will change, I will provide the output once we have it.

Reviewer 3: We can merge as to not hold this up, then update the output later.

Reviewer 4: If we merge this way let's at least update to 
```suggestion
kubectl get providers.clusterctl.cluster.x-k8s.io --all-namespaces
```

We have at least two other CRDs called providers:
```
$ k api-resources | grep -i provider
providers                                                                      clusterctl.cluster.x-k8s.io/v1alpha3                           true         Provider
providers                                                                      externaldata.gatekeeper.sh/v1alpha1                            false        Provider
providers                                                                      notification.toolkit.fluxcd.io/v1beta1                         true         Provider

```

cappp-system                        infrastructure-preprovisioned   5m44s   InfrastructureProvider   preprovisioned   v0.5.0
capv-system                         infrastructure-vsphere          57s     InfrastructureProvider   vsphere          v1.1.1
capz-system                         infrastructure-azure            5m47s   InfrastructureProvider   azure                 v1.1.1
```
If the upgrade fails, review the prerequisites section and ensure that you've followed the steps in the [DKP upgrade overview][dkpup].

## Upgrade the core addons

To install the core addons, DKP relies on the `ClusterResourceSet` [Cluster API feature][CAPI]. In the CAPI component upgrade, we deleted the previous set of outdated global `ClusterResourceSets` because prior to DKP 2.2 some addons were installed using a global configuration. In order to support individual cluster upgrades, DKP 2.2 now installs all addons with a unique set of `ClusterResourceSet` and corresponding referenced resources, all named using the cluster’s name as a suffix. For example: `calico-cni-installation-my-aws-cluster`.

> <p class="message--warning"><strong>WARNING:</strong> If you have modified any of the <code>clusterResourceSet</code> definitions, these changes will <strong>not</strong> be preserved when running the command <code>dkp upgrade addons</code>. You can use the <code>--dry-run -o yaml</code> options to save the new configuration to a file and remake the same changes upon each upgrade.</p>

Reviewer: ```suggestion
<p class="message--warning"><strong>WARNING:</strong> If you have modified any of the <code>clusterResourceSet</code> definitions, these changes will <strong>not</strong> be preserved when running the command <code>dkp upgrade addons</code>. You must use the <code>--dry-run -o yaml</code> options to save the new configuration to a file and remake the same changes upon each upgrade.</p>
```

> Your cluster comes preconfigured with a few different core addons that provide functionality to your cluster upon creation. These include: CSI, CNI, Cluster Autoscaler, and Node Feature Discovery. New versions of DKP may come prebundled with newer versions of these addons. Perform the following steps to update these addons.

Reviewer: Should we mention somewhere that this needs to run for each managed and management cluster with the `KUBECONFIG` configured for each cluster?

Author: Yes, i'll mention that in the Overview section beneath the stepwise instructions:

> Run all three steps on the management cluster (Kommander cluster) first. Then, run the second and third step on additional managed clusters (Konvoy clusters), one cluster at a time with the KUBECONFIG configured for each cluster.

configmap/calico-cni-installation-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/tigera-operator-my-aws-cluster upgraded
configmap/tigera-operator-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/aws-ebs-csi-my-aws-cluster upgraded
configmap/aws-ebs-csi-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/cluster-autoscaler-my-aws-cluster upgraded
configmap/cluster-autoscaler-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/node-feature-discovery-my-aws-cluster upgraded
configmap/node-feature-discovery-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/nvidia-feature-discovery-my-aws-cluster upgraded
configmap/nvidia-feature-discovery-my-aws-cluster upgraded
```

1. Monitor the pods for the core addons restarting in your cluster:

```bash
###DEV ADD OUTPUT HERE
```
Once complete, begin upgrading the Kubernetes version.

## Upgrade the Kubernetes version

When upgrading the Kubernetes version of a cluster, first upgrade the control plane and then the node pools.

1. Replace `my-aws-cluster` with the name of the cluster.

> 1. Use the CLUSTERNAME flag to reference the cluster on which you want to upgrade the Kubernetes version.

Reviewer: ```suggestion
1. Use the `--cluster-name` flag to reference the cluster on which you want to upgrade the Kubernetes version.
```

1. Upgrade the Kubernetes version of the control plane.

```bash
dkp update controlplane aws --cluster-name=${CLUSTER_NAME} --kubernetes-version=v1.22.8
```

The output should be similar to:

```bash
### DEV ENTER OUTPUT HERE
```

1. Upgrade the Kubernetes version of each of your node pools. Replace `my-nodepool` with the name of the node pool.

```bash
export NODEPOOL_NAME=my-nodepool
dkp update nodepool aws ${NODEPOOL_NAME} --cluster-name=${CLUSTER_NAME} --kubernetes-version=v1.22.8
```
The output should be similar to:

```bash
### DEV ENTER OUTPUT HERE
```
Repeat this step for each additional node pool.

> 1. If you have any additional management or managed Clusters, review the [DKP Upgrade][dkpup] documentation.

Reviewer 2: ```suggestion
1. If you have any additional management or managed clusters, review the [DKP Upgrade][dkpup] documentation for next steps.
```

[dkup]: .../dkp-upgrade/
[upgradekomm]: .../upgrade-kommander/
[supportsite]: https://support.d2iq.com/hc/en-us
[airgapbundle]: .../konvoy/2.2/choose-infrastructure/airgapbundle/
[dkpenterprise]: .../kommander/2.2/licensing/enterprise/
[kubeconfig]: https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/

> [CAPI]: https://cluster-api.sigs.k8s.io/

Reviewer 2: ```suggestion
[CAPI]: https://cluster-api.sigs.k8s.io/
[releasenotes]: ../../../release-notes
```

---
layout: layout.pug
navigationTitle: DKP Enterprise
title: Upgrade Konvoy for DKP Enterprise
excerpt: Upgrade your Konvoy environment within the DKP Enterprise license.
beta: false
enterprise: false
menuWeight: 30
---

## Prerequisites

> * Follow the steps listed in the [DKP upgrade overview][dkpup].

Reviewer 2: ```suggestion
* **REQUIRED** Before upgrading, create an [on-demand backup][backup] of your current configuration with Velero.

* Follow the steps listed in the [DKP upgrade overview][dkpup].
```

3. Upgrade the Kubernetes version of the control plane.

```bash
dkp update controlplane aws --cluster-name=${CLUSTER_NAME} --kubernetes-version=v1.22.8
```

The output should be similar to:

```bash
### DEV ENTER OUTPUT HERE
```

4. Upgrade the Kubernetes version of each of your node pools. Replace `my-nodepool` with the name of the node pool.

```bash
export NODEPOOL_NAME=my-nodepool
dkp update nodepool aws ${NODEPOOL_NAME} --cluster-name=${CLUSTER_NAME} --kubernetes-version=v1.22.8
```
The output should be similar to:

```bash
### DEV ENTER OUTPUT HERE
```
Repeat this step for each additional node pool.

If you have any additional management or managed clusters, review the [DKP Upgrade][dkpup] documentation for next steps.

[dkpup]: ../../dkp-upgrade/
[upgradekomm]: ../../upgrade-kommander/
[supportsite]: https://support.d2iq.com/hc/en-us
[airgapbundle]: /dkp//konvoy/2.2/choose-infrastructure/airgapbundle/
[dkpenterprise]: /dkp/kommander/2.2/licensing/enterprise/
[kubeconfig]: https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/
[CAPI]: https://cluster-api.sigs.k8s.io/
[releasenotes]: ../../../release-notes

> [envariables]: /dkp/konvoy/2.2/choose-infrastructure/azure/quick-start-azure/#configure-azure-prerequisites

Reviewer 2: ```suggestion
[envariables]: /dkp/konvoy/2.2/choose-infrastructure/azure/quick-start-azure/#configure-azure-prerequisites
[backup]: ../backup-and-restore/#back-up-on-demand
```

✓ Initializing new CAPI components
✓ Deleting Outdated Global ClusterResourceSets
```

If the upgrade fails, review the prerequisites section and ensure that you've followed the steps in the [DKP upgrade overview][dkpup].

## Upgrade the core addons

To install the core addons, DKP relies on the `ClusterResourceSet` [Cluster API feature][CAPI]. In the CAPI component upgrade, we deleted the previous set of outdated global `ClusterResourceSets` because prior to DKP 2.2 some addons were installed using a global configuration. In order to support individual cluster upgrades, DKP 2.2 now installs all addons with a unique set of `ClusterResourceSet` and corresponding referenced resources, all named using the cluster’s name as a suffix. For example: `calico-cni-installation-my-aws-cluster`.

<p class="message--warning"><strong>WARNING:</strong> If you have modified any of the <code>clusterResourceSet</code> definitions, these changes will <strong>not</strong> be preserved when running the command <code>dkp upgrade addons</code>. You must use the <code>--dry-run -o yaml</code> options to save the new configuration to a file and remake the same changes upon each upgrade.</p>

> Your cluster comes preconfigured with a few different core addons that provide functionality to your cluster upon creation. These include: CSI, CNI, Cluster Autoscaler, and Node Feature Discovery. New versions of DKP may come prebundled with newer versions of these addons. Perform the following steps to update these addons.

Reviewer 5: ```suggestion
Your cluster comes preconfigured with a few different core addons that provide functionality to your cluster upon creation. These include: CSI, CNI, Cluster Autoscaler, and Node Feature Discovery. New versions of DKP may come pre-bundled with newer versions of these addons. Perform the following steps to update these addons.
```
```suggestion
Your cluster comes preconfigured with a few different core addons that provide functionality to your cluster upon creation. These include: CSI, CNI, Cluster Autoscaler, and Node Feature Discovery. New versions of DKP may come prebundled with newer versions of these addons. Perform the following steps to update these addons.
```

Your cluster comes preconfigured with a few different core addons that provide functionality to your cluster upon creation. These include: CSI, CNI, Cluster Autoscaler, and Node Feature Discovery. New versions of DKP may come prebundled with newer versions of these addons. Perform the following steps to update these addons.

<p class="message--warning"><strong>IMPORTANT:</strong>Ensure your <code>dkp</code> configuration references the management cluster where you want to run the upgrade by setting the <code>KUBECONFIG</code> environment variable <a href="https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/">to the appropriate kubeconfig file location</a>.</p>

An alternative to initializing the KUBECONFIG environment variable is to use the `--kubeconfig=cluster_name.conf flag`.

1. Replace `my-aws-cluster` with the name of the cluster.

```bash
export CLUSTER_NAME=my-aws-cluster
dkp upgrade addons aws --cluster-name=${CLUSTER_NAME}
```

The output should be similar to:

> ```bash

Reviewer 6: ```suggestion
```text
```

Generating addon resources

> clusterresourceset.addons.cluster.x-k8s.io/calico-cni-    installation-my-aws-cluster upgraded

Reviewer 6: ```suggestion
clusterresourceset.addons.cluster.x-k8s.io/calico-cni-installation-my-aws-cluster upgraded
```

configmap/calico-cni-installation-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/tigera-operator-my-aws-cluster upgraded
configmap/tigera-operator-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/aws-ebs-csi-my-aws-cluster upgraded
configmap/aws-ebs-csi-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/cluster-autoscaler-my-aws-cluster upgraded
configmap/cluster-autoscaler-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/node-feature-discovery-my-aws-cluster upgraded
configmap/node-feature-discovery-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/nvidia-feature-discovery-my-aws-cluster upgraded

> configmap/nvidia-feature-discovery-my-aws-cluster upgraded

Reviewer 6: ```suggestion
Generating addon resources
Preparing cluster for upgrade to snapshot.storage.k8s.io/v1.
Generating new CRS configuration.
clusterresourceset.addons.cluster.x-k8s.io/calico-cni-installation-my-aws-cluster upgraded
configmap/calico-cni-installation-my-aws-cluster upgraded
configmap/tigera-operator-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/aws-ebs-csi-my-aws-cluster upgraded
configmap/aws-ebs-csi-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/cluster-autoscaler-my-aws-cluster upgraded
configmap/cluster-autoscaler-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/node-feature-discovery-my-aws-cluster upgraded
configmap/node-feature-discovery-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/nvidia-feature-discovery-my-aws-cluster upgraded
configmap/nvidia-feature-discovery-my-aws-cluster upgraded
Deleting resources created to prepare cluster for upgrade to snapshot.storage.k8s.io/v1.
 ✓ Upgrading the core addons
```
```

> ```

Reviewer 6: ```text
Generating addon resources
Preparing cluster for upgrade to snapshot.storage.k8s.io/v1.
Generating new CRS configuration.
clusterresourceset.addons.cluster.x-k8s.io/calico-cni-installation-my-aws-cluster upgraded
configmap/calico-cni-installation-my-aws-cluster upgraded
configmap/tigera-operator-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/aws-ebs-csi-my-aws-cluster upgraded
configmap/aws-ebs-csi-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/cluster-autoscaler-my-aws-cluster upgraded
configmap/cluster-autoscaler-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/node-feature-discovery-my-aws-cluster upgraded
configmap/node-feature-discovery-my-aws-cluster upgraded
clusterresourceset.addons.cluster.x-k8s.io/nvidia-feature-discovery-my-aws-cluster upgraded
configmap/nvidia-feature-discovery-my-aws-cluster upgraded
Deleting resources created to prepare cluster for upgrade to snapshot.storage.k8s.io/v1.
 ✓ Upgrading the core addons
```

2. Monitor the pods for the core addons restarting in your cluster:

```bash
kubectl rollout status daemonset/calico-node
--namespace calico-system
```

The output should be similar to:

```bash
daemon set "calico-node" successfully rolled out

> ```

Reviewer 6: This step requires the user to interact with the _managed_ cluster, whereas all other steps in this doc require the user to interact with the _management_ cluster.

Let's remove Step 2 for now. If we come up with an alternative, we'll add it here.

Once complete, begin upgrading the Kubernetes version.

## Upgrade the Kubernetes version

When upgrading the Kubernetes version of a cluster, first upgrade the control plane and then the node pools.

1. Replace `my-aws-cluster` with the name of the cluster.

2. Use the `--cluster-name` flag to reference the cluster on which you want to upgrade the Kubernetes version.

3. Upgrade the Kubernetes version of the control plane.

```bash
dkp update controlplane aws --cluster-name=${CLUSTER_NAME} --kubernetes-version=v1.22.8
```

The output should be similar to:

> ```bash

Reviewer 6: ```suggestion
```text
```

> ### DEV ENTER OUTPUT HERE

Reviewer 6: ```suggestion
Updating control plane resource controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane default/my-aws-cluster-control-plane
Waiting for control plane update to finish.
 ✓ Updating the control plane ```

> ```

Reviewer 6: ```text
Updating control plane resource controlplane.cluster.x-k8s.io/v1beta1, Kind=KubeadmControlPlane default/my-aws-cluster-control-plane
Waiting for control plane update to finish.
 ✓ Updating the control plane 
```

4. Upgrade the Kubernetes version of each of your node pools. Replace `my-nodepool` with the name of the node pool.

```bash
export NODEPOOL_NAME=my-nodepool
dkp update nodepool aws ${NODEPOOL_NAME} --cluster-name=${CLUSTER_NAME} --kubernetes-version=v1.22.8
```
The output should be similar to:

> ```bash

Reviewer 6: ```suggestion
```text
```

> ### DEV ENTER OUTPUT HERE

Reviewer 6: ```suggestion
Updating node pool resource cluster.x-k8s.io/v1beta1, Kind=MachineDeployment default/my-aws-cluster-my-nodepool
Waiting for node pool update to finish.
 ✓ Updating the my-aws-cluster-my-nodepool node pool 
```

Run all three steps on the management cluster (Kommander cluster) first. Then, run the second and third steps on additional managed clusters (Konvoy clusters), one cluster at a time using the KUBECONFIG configured for each cluster. If an AMI was specified when initially creating a cluster, you must build a new one with [Konvoy Image Builder][KIB] and pass it with `--ami.`

<p class="message--note"><strong>NOTE:</strong> For pre-provisioned air-gapped environments, you must run <code>konvoy-image upload artifacts</code>.</p>

For a full list of DKP Enterprise features, see [DKP Enterprise][dkpenterprise].

<p class="message--note"><strong>NOTE:</strong> You must maintain your attached clusters manually. Review the documentation from your cloud provider for more information.</p>

## Upgrade the CAPI components

New versions of DKP come pre-bundled with newer versions of CAPI, newer versions of infrastructure providers, or new infrastructure providers. When using a new version of the DKP CLI, upgrade all of these components first.

If you are running on more than one management cluster (Kommander cluster), you must upgrade the CAPI components on each of these clusters.

> <p class="message--warning"><strong>IMPORTANT:</strong>Ensure your <code>dkp</code> configuration references the management cluster where you want to run the upgrade by setting the <code>KUBECONFIG</code> environment variable <a href="https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/">to the appropriate kubeconfig file location</a>.</p>
> 
> An alternative to initializing the KUBECONFIG environment variable is to use the <code>--kubeconfig=cluster_name.conf</code> flag.

Reviewer 6: I suggest combining these into one note:

```suggestion
<p class="message--warning"><strong>IMPORTANT:</strong>Ensure your <code>dkp</code> configuration references the management cluster where you want to run the upgrade by setting the <code>KUBECONFIG</code> environment variable, or using the <code>--kubeconfig</code> flag, <a href="https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/">in accordance with Kubernetes conventions</a>.
```

I also noticed that the "<code>" tag doesn't seem to render monospaced in the note, as I've underlined here: 
![image](https://user-images.githubusercontent.com/3445370/161829507-487beba3-b6a2-49f8-83e2-480d3fcc7ebb.png)

Author: I'll go ahead and commit this as well to the Essential doc. I just didn't want a huge note block, but the two concepts are related.

1. Run the following upgrade command for the CAPI components.

```bash
dkp upgrade capi-components
```

The output resembles the following:

> ```bash

Reviewer 6: Most--if not all--output blocks should use the "text" syntax, _not_ shell, because the output should not be highlighted as a shell script would. For example, "for" should not be highlighted blue here:
![image](https://user-images.githubusercontent.com/3445370/161829931-7383d479-2782-4908-a5f9-2465bdff9090.png)

```suggestion
```text
```

Author: This is good insight, thank you.

Reviewer 5: Also, for future reference @mwrightd2 @dlipovetsky 

````
```bash
````

should be reserved for commands - because it keeps the copy/paste icon.

````
```sh
````

Will remove that in styling. So, `sh` would be a good option here, too.
